streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -526,15 +555,30 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
 
         // ---------------- Step Two ---------------- //
 
-        // assign tasks to clients
         final Map<UUID, ClientState> states = new HashMap<>();
         for (final Map.Entry<UUID, ClientMetadata> entry : clientMetadataMap.entrySet()) {
-            states.put(entry.getKey(), entry.getValue().state);
+            final ClientState state = entry.getValue().state;
+            states.put(entry.getKey(), state);
+
+            // Either the active tasks (eager) OR the owned partitions (cooperative) were encoded in the subscription
+            // according to the rebalancing protocol, so convert any partitions in a client to tasks where necessary
+            if (!state.ownedPartitions().isEmpty()) {
```

Comment:
Yes, `ownedPartitions` should only be non-empty when following cooperative protocol. Note though that in the Streams upgrade path, some clients may be on eager while some are on cooperative, so we must check all `ownedPartitions` even if the leader is on eager.
The reason for doing this is because the ownedPartitions should be the ultimate "source of truth" for a consumer, for a few reasons: the task <-> partitions mapping could change, or a StreamThread could have failed to initialize a task and thus not included it in the encoded `activeTasks`, but the consumer still owns the task and we can't give it away until it's been revoked

Reply:
During assignment we go by the `ownedPartitions` as the ultimate source of truth, w.r.t which consumer to try and give tasks back to and which tasks should be removed from the assignment because they were previously owned. If a consumer on eager claims task1 while a cooperative consumer claims partition1, we will ultimately try and give task1/partition1 back to the cooperative consumer. And if it has to be assigned to another consumer, even the eager consumer that claimed task1, we will still remove it from the assignment until the cooperative consumer can revoked it.

We also check that the task <-> partitions mapping does not change such that (for example) task1 maps to partition1 and partition2 which are claimed by different consumers.

Reply:
Does that make sense? In the end we always go with the "safest" assignment and don't assign partitions claimed in someone's ownedPartitions to anyone else. If the scenario you described occurs. Which I believe is possible, if for example the eager consumer misses a rebalance it won't have cleared it's prevActiveTasks. However we will still try and give it back to the cooperative consumer, and if not to them then to no one (during this rebalance).

Reply:
SG, thanks for the explanation!